src/node_state.rs

```diff
@@ -86,37 +89,52 @@ impl NodeState {
 
     fn handle_proof(&mut self, tx: ProofTransaction) -> Result<(), Error> {
         // Diverse verifications
-        let unsettled_tx = self
-            .unsettled_transactions
-            .get(&tx.blob_tx_hash)
-            .context("Tx is either settled or does not exists.")?;
+        let unsettled_tx: Vec<&UnsettledTransaction> = tx
+            .blobs_references
+            .iter()
+            .map(|blob_ref| self.unsettled_transactions.get(&blob_ref.blob_tx_hash))
+            .collect::<Vec<Option<&UnsettledTransaction>>>()
+            .sequence()
+            .context("At lease 1 tx is either settled or does not exists")?;
 
         // Verify proof
-        let blob_detail = Self::verify_proof(&tx)?;
+        let blobs_detail = Self::verify_proof(&tx)?;
 
         // hash payloads
-        let blobs_hash = Self::extract_blobs_hash(&blob_detail)?;
+        let extracted_blobs_hash = Self::extract_blobs_hash(&blobs_detail)?;
 
+        let initial_blobs_hash = unsettled_tx
+            .iter()
+            .map(|tx| tx.blobs_hash.clone())
+            .collect::<Vec<BlobsHash>>();
         // some verifications
-        if blobs_hash != unsettled_tx.blobs_hash {
+        if extracted_blobs_hash != initial_blobs_hash {
             bail!(
-                "Proof blobs hash '{}' do not correspond to transaction blobs hash '{}'.",
-                blobs_hash,
-                unsettled_tx.blobs_hash
+                "Proof blobs hash '{:?}' do not correspond to transaction blobs hash '{:?}'.",
+                extracted_blobs_hash,
+                initial_blobs_hash
             )
         }
 
-        self.save_blob_details(&tx, blob_detail)?;
+        self.save_blob_details(&tx, blobs_detail)?;
 
-        let is_next_to_settle = self
-            .unsettled_transactions
-            .is_next_unsettled_tx(&tx.blob_tx_hash, &tx.contract_name);
-
-        // check if tx can be settled
-        if is_next_to_settle && self.is_ready_for_settlement(&tx.blob_tx_hash) {
-            // settle tx
-            self.settle_tx(&tx)?;
-        }
+        tx.blobs_references
+            .iter()
+            .map(|blob_ref| {
+                let is_next_to_settle = self
+                    .unsettled_transactions
+                    .is_next_unsettled_tx(&blob_ref.blob_tx_hash, &blob_ref.contract_name);
```

Comment:
En principe tu l'as dans blobs_detail l'info du contract name non ?

Reply:
oui, en soit en effet je peux retrouver l'info autrement, mais ça complexifie encore l'implémentation car on est en train d'itérer sur un vec, et pour trouver le contract name qui correspond à cet instant, il faut itérer sur le vec de blobs_detail pour trouver un Option<ContractName>... ça fait beaucoup d'imbrication, de cas d'erreur à gérer (sauf à faire des unwrap() mais faut pas faire ça), alors qu'il est extrêmement plus simple de l'avoir dans le blob_ref... 

ça ne me semble pas très coûteux en bande passante que de l'avoir là même si c'est un peu redondant, en attendant ça simplifie sensiblement l'implémentation...

Reply:
Ça me va si ça complexifie franchement le code